Jint/Pooling/EnginePool.cs

```diff
@@ -0,0 +1,110 @@
+using System;
+using Esprima;
+using Esprima.Ast;
+using Jint.Native;
+
+namespace Jint.Pooling
+{
+    /// <summary>
+    /// Offers a pool of engines that can be initialized to certain state and then used with temporary state that
+    /// will be discarded.
+    /// </summary>
+    public sealed class EnginePool
+    {
+        private readonly ObjectPool<Engine> _enginePool;
+
+        public EnginePool(int size, Options options, Action<IPooledEngine> initialize)
+        {
+            _enginePool = new ObjectPool<Engine>(factory: () =>
+            {
+                var engine = new Engine(options);
+                initialize?.Invoke(new EngineWrapper(engine));
+                return engine;
+            }, size);
+        }
+
+        public void ExecuteInIsolatedContext(Action<IPooledEngine> action)
+        {
+            Engine engine;
+            lock (_enginePool)
+            {
+                engine = _enginePool.Allocate();
+            }
+
+            using (engine.EnterIsolatedContext())
+            {
+                action(new EngineWrapper(engine));
+            }
+
+            lock (_enginePool)
+            {
+                _enginePool.Free(engine);
+            }
+        }
+
+        private sealed class EngineWrapper : IPooledEngine
```

Comment:
Maybe create an `IEngine` instead that is implemented by `Engine` too. And the pooled version could handle the dispose pattern, and have a Release/Return() method if we don't want to use the dispose pattern.

Reply:
Can do, I just want to IEngine to strictly limited to execution and setting values to context. Might be role interface too. Just need to limit public access to things like `ObjectConstructor` etc which are now public and might allow doing bad things.

Reply:
I think I'll stick for wrapper now so that we can implement the dispose and checks here. Or should the Engine implement it as quite no-op for now and we would just start checking `if (_disposed) { throw diposedexception}` in engine service methods?